src/allotropy/parsers/thermo_fisher_nanodrop_eight/nanodrop_eight_parser.py

```diff
@@ -0,0 +1,271 @@
+from typing import Optional, TypeVar
+import uuid
+
+import pandas as pd
+
+from allotropy.allotrope.models.shared.definitions.custom import (
+    TQuantityValueMicrogramPerMicroliter,
+    TQuantityValueMicrogramPerMilliliter,
+    TQuantityValueMilliAbsorbanceUnit,
+    TQuantityValueMilligramPerMilliliter,
+    TQuantityValueNanogramPerMicroliter,
+    TQuantityValueNanogramPerMilliliter,
+    TQuantityValueNanometer,
+    TQuantityValuePicogramPerMilliliter,
+)
+from allotropy.allotrope.models.shared.definitions.definitions import (
+    TQuantityValue,
+    ValueEnum,
+)
+from allotropy.allotrope.models.spectrophotometry_benchling_2023_12_spectrophotometry import (
+    CalculatedDataAggregateDocument,
+    CalculatedDataDocumentItem,
+    DataSourceAggregateDocument1,
+    DataSourceDocumentItem,
+    DataSystemDocument,
+    DeviceSystemDocument,
+    MeasurementAggregateDocument,
+    Model,
+    SampleDocument,
+    SpectrophotometryAggregateDocument,
+    SpectrophotometryDocumentItem,
+    UltravioletAbsorbancePointDetectionDeviceControlAggregateDocument,
+    UltravioletAbsorbancePointDetectionDeviceControlDocumentItem,
+    UltravioletAbsorbancePointDetectionMeasurementDocumentItems,
+)
+from allotropy.constants import ASM_CONVERTER_NAME, ASM_CONVERTER_VERSION
+from allotropy.named_file_contents import NamedFileContents
+from allotropy.parsers.thermo_fisher_nanodrop_eight.nanodrop_eight_reader import (
+    NanoDropEightReader,
+)
+from allotropy.parsers.vendor_parser import VendorParser
+
+T = TypeVar("T")
+
+CONCENTRATION_UNIT_TO_TQUANTITY = {
+    "ug/ul": TQuantityValueMicrogramPerMicroliter,
+    "ug/ml": TQuantityValueMicrogramPerMilliliter,
+    "mg/ml": TQuantityValueMilligramPerMilliliter,
+    "ng/ul": TQuantityValueNanogramPerMicroliter,
+    "ng/ml": TQuantityValueNanogramPerMilliliter,
+    "pg/ul": TQuantityValuePicogramPerMilliliter,
+}
+
+
+def _get_str(data_frame: pd.DataFrame, row: int, column: str) -> Optional[str]:
+    if column not in data_frame.columns:
+        return None
+    if pd.isna(data_frame.iloc[row][column]):
+        return None
+
+    return str(data_frame.iloc[row][column])
+
+
+def _get_float(data_frame: pd.DataFrame, row: int, column: str) -> float | ValueEnum:
+    try:
+        return float(data_frame.iloc[row][column])
+    except (ValueError, TypeError):
+        return ValueEnum.NaN
+
+
+def _get_concentration(conc: float | ValueEnum, unit: Optional[str]) -> Optional[T]:
+    if unit in CONCENTRATION_UNIT_TO_TQUANTITY and isinstance(conc, float):
+        cls = CONCENTRATION_UNIT_TO_TQUANTITY[unit]
+        return cls(value=conc)  # type: ignore[return-value]
```

Comment:
We should be able to remove this ignore. I'll get back to you on this.

Reply:
Eh, it's not going to be possible to remove this suppression without some larger refactoring of type classes. I'm fine leaving this in for now.

Reply:
I figured out how to do this. It does require some ugly union types though. I think I'll just push my changes.

Reply:
er, that's not a good operating model. So here's what you want:
1. Delete the declaration of `T` -- generics aren't helpful here.
2. Add this block above:
```
ConcentrationType = Union[
    TQuantityValueMicrogramPerMicroliter,
    TQuantityValueMicrogramPerMilliliter,
    TQuantityValueMilligramPerMilliliter,
    TQuantityValueNanogramPerMicroliter,
    TQuantityValueNanogramPerMilliliter,
    TQuantityValuePicogramPerMilliliter,
]
ConcentrationClassType = Union[
    type[TQuantityValueMicrogramPerMicroliter],
    type[TQuantityValueMicrogramPerMilliliter],
    type[TQuantityValueMilligramPerMilliliter],
    type[TQuantityValueNanogramPerMicroliter],
    type[TQuantityValueNanogramPerMilliliter],
    type[TQuantityValuePicogramPerMilliliter],
]
```
3. Add a type hint:
```
CONCENTRATION_UNIT_TO_TQUANTITY: Mapping[str, ConcentrationClassType] = {
``` 

4. This method should be:
```
def _get_concentration(
    conc: JsonFloat, unit: Optional[str]
) -> Optional[ConcentrationType]:
    if unit and unit in CONCENTRATION_UNIT_TO_TQUANTITY and isinstance(conc, float):
        cls = CONCENTRATION_UNIT_TO_TQUANTITY[unit]
        return cls(value=conc)

    return None
```